### PR TITLE
CHAIN-99: drive websocket broadcasts from DB notifies

### PR DESCRIPTION
**Pull request by bflood366**

### CODE REVIEW

File: backend/src/main/kotlin/co/chainring/core/utils/PgListener.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                                try {
+                                    onNotifyLogic(it)
+                                } catch (e: Exception) {
+                                    logger.error(e) { "Exception processing notification" }
+                                }
+                            }
+                        }
+                    }
+                } catch (_: InterruptedException) {
+                    listenerConn.createStatement().also { it.execute("UNLISTEN $channel") }.close()
```
Comment (bholzman) [inline backend/src/main/kotlin/co/chainring/core/utils/PgListener.kt:47]: Should we catch any errors here?

**Review by bholzman — APPROVED** [review]
Looks great!

File: backend/src/main/kotlin/co/chainring/core/utils/PgListener.kt (hunk 1 above)
Comment (bflood366) [reply backend/src/main/kotlin/co/chainring/core/utils/PgListener.kt]: I added a try / catch around the unlisten/close

File: backend/src/main/kotlin/co/chainring/core/websocket/Broadcaster.kt [hunk 2]
```diff
@@ -317,22 +418,40 @@ class Broadcaster {
...
+                }
+            }
+        } catch (e: Exception) {
+            logger.error(e) { "Broadcaster: Unhandled exception " }
         }
+    }
 
-        findClients(principal, topic)
-            .forEach { it.send(OutgoingWSMessage.Publish(topic, message)) }
+    fun dbNotify(notifications: List<PrincipalNotifications>) {
```
[the remark this answers is not in the document]
Comment (bflood366) [reply backend/src/main/kotlin/co/chainring/core/websocket/Broadcaster.kt]: ah good idea, made this change and removed injecting `broadcaster` into `ExchangeService`

**Pull request merged**